Fix primtorch out_wrapper semantics for factory functions

soulitzer opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #82284
* __->__ #82375

This PR:
- introduces new OpInfo attribute `is_factory_function`
- updates OpInfo test_out to handle case when `is_factory_function=True`:
- correct primtorch out_wrapper
- update sample inputs for arange, linspace, logspace to not explicitly pass in dtype or device (having this sample is necessary for the test to get triggered)


Fixes https://github.com/pytorch/pytorch/issues/82364

soulitzer committed:
Fix primtorch out_wrapper semantics for factory functions

[ghstack-poisoned]

soulitzer committed:
Update on "Fix primtorch out_wrapper semantics for factory functions"




[ghstack-poisoned]

soulitzer commented on torch/testing/_internal/common_methods_invocations.py:
```diff
@@ -797,6 +797,8 @@ class OpInfo(object):
 
     supports_expanded_weight: bool = False
 
+    is_factory_function: bool = False
```
Maybe there is a better way to infer that its a factory function?

ezyang replied:
this seems about as good as it gets

ezyang commented:
While this certainly fixes the problem, it would be nice to get a comment from @ngimel about whether or not the original behavior is intended, or a bug that shouldn't be replicated. In particular, it seems like `safe_copy_out` was added *explicitly* to prevent this sort of automatic dtype/device conversion. If it's not supposed to be a bug, we might want to just structure the PrimTorch ref so that it follows the original implementation more correctly. If it is a bug, we should probably fix up the OpInfo not to exercise the buggy case (and put the buggy version under a deprecation cycle.)

It does seem like it's intentional though; other factory out variants have the same behavior

```
>>> torch.ones(3, out=torch.empty(3, dtype=torch.long))
tensor([1, 1, 1])```

ezyang commented on torch/_prims_common/wrappers.py:
```diff
@@ -191,8 +191,18 @@ def _out_wrapper(fn: Callable) -> Callable:
         )
 
+        sig = inspect.signature(fn)
+        factory_kwargs = ("device", "dtype")
```
This doesn't seem sufficient; it seems layout is also propagated:

```
>>> torch.zeros(3, out=torch.empty(3, dtype=torch.long, layout=torch.sparse_coo))
tensor(indices=tensor([], size=(1, 0)),
       values=tensor([], size=(0,)),
       size=(3,), nnz=0, layout=torch.sparse_coo)
```

soulitzer replied:
Yup, this would need to be updated - leaving this as a follow up as most factory functions implemented as refs only support the dtype and device kwargs

ezyang commented on torch/_prims_common/wrappers.py:
```diff
@@ -191,8 +191,18 @@ def _out_wrapper(fn: Callable) -> Callable:
 
+        sig = inspect.signature(fn)
+        factory_kwargs = ("device", "dtype")
+        is_factory_fn = all(p in sig.parameters for p in factory_kwargs)
```
I audited `native_functions.yaml` to check that this wouldn't fire inappropriately, and it appears to pass the audit, though I'm still a little nervous

soulitzer replied:
Thanks for checking - yeah I guess our assumption here is that any function that takes factory kwargs and accepts out arg should infer factory kwargs from out args if the factory kwargs are not specified. Improving our OpInfo (and sample input) coverage should help with checking this as well.

soulitzer committed:
Update on "Fix primtorch out_wrapper semantics for factory functions"


This PR:
- introduces new OpInfo attribute `is_factory_function`
- updates OpInfo test_out to handle case when `is_factory_function=True`:
- correct primtorch out_wrapper
- update sample inputs for arange, linspace, logspace to not explicitly pass in dtype or device (having this sample is necessary for the test to get triggered)


Fixes https://github.com/pytorch/pytorch/issues/82364


[ghstack-poisoned]

soulitzer commented:
@pytorchbot merge -g
